File: 알고리즘 문제풀이/kakao/뉴스클러스터링.py
#J(A,B): A,B의 교집합 크기를 두 집함의 합집합 크기로 나눈 값으로 정의
#ex) A={1,2,3}, B={2,3,4} J(A,B)=2/4=0.5
#A,B 모두 공집합인 경우 J(A,B)=1
#A,B는 다중집합,
#A ∩ B는 원소 1을 min(3, 5)인 3개
#A ∪ B는 원소 1을 max(3, 5)인 5개 
from collections import defaultdict
def f1(d1,d2):#합집합,크기 반환  (d={**d1,**d2})
    d=defaultdict(int)
    for key in d1.keys():
        if key in d2.keys():
            d[key]=max(d1[key],d2[key])
        else:
            d[key]=d1[key]
    for key in d2.keys():
        if key not in d1.keys():
            d[key]=d2[key]
    cnt=0
    for val in d.values():
         cnt+=val   
    return cnt
def f2(d1,d2):#교집합,크기 반환
    d=defaultdict(int)
    for key in d1.keys():
        if key in d2.keys():
            d[key]=min(d1[key],d2[key])
    cnt=0
    for val in d.values():
        cnt+=val
    return cnt
def solution(str1, str2):
    d1,d2=defaultdict(int),defaultdict(int)
    for i in range(0,len(str1)-1):
        tmp=str1[i:i+2].upper()
        if tmp.isalpha():
            d1[tmp]+=1
    for i in range(0,len(str2)-1):
        tmp=str2[i:i+2].upper()
        if tmp.isalpha():
            d2[tmp]+=1
    c1=f1(d1,d2) #합집합 크기
    c2=f2(d1,d2) #교집합 크기
    if c1==0 and c2==0: #공집합인 경우
        return 65536
    return int((c2/c1)*65536)

File: 알고리즘 문제풀이/Algorithm/kakao/뉴스클러스터링.js
function getSet(s) {
  let ss = s,
    arr = [];
  while (ss.length > 0) {
    const word = ss.slice(0, 2);
    if (word.length < 2) break;
    if (word.search(/[a-z][a-z]/) !== -1) arr.push(word);
    ss = ss.slice(1);
  }
  return arr;
}
function interSection(a, b) {
  let result = 0;
  Object.keys(a).forEach((key) => {
    if (b[key] !== undefined) result += Math.min(a[key], b[key]);
    else result += a[key];
  });
  return result;
}
function union(a, b) {
  let result = 0,
    tmpA = a,
    tmpB = b;
  Object.keys(tmpB).forEach((key) => {
    if (tmpA[key] !== undefined) tmpA = Math.max(tmpA[key], tmpB[key]);
    else tmpA[key] = tmpB[key];
  });
  Object.keys(tmpA).forEach((key) => {
    result += tmpA[key];
  });
  return result;
}
function solution(str1, str2) {
  const a = getSet(str1),
    b = getSet(str2);
  if (a.lenght === 0 && b.length === 0) return 65536;
  else {
    const mi = interSection(a, b);
    ma = union(a, b);
    return parseInt((mi / ma) * 65536, 10);
  }
}
